centchain/api.go

```diff
@@ -0,0 +1,94 @@
+package centchain
+
+import (
+	gsrpc "github.com/centrifuge/go-substrate-rpc-client"
+	"github.com/centrifuge/go-substrate-rpc-client/client"
+	"github.com/centrifuge/go-substrate-rpc-client/rpc/author"
+	"github.com/centrifuge/go-substrate-rpc-client/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/types"
+)
+
+// API exposes required functions to interact with Centrifuge Chain.
+type API interface {
+	// GetMetadataLatest returns latest metadata from the centrifuge chain.
+	GetMetadataLatest() (*types.Metadata, error)
+
+	// SubmitExtrinsic signs the given call with the provided KeyRingPair and submits an extrinsic.
+	// Returns transaction hash, latest block number before extrinsic submission, and signature attached with the extrinsic.
+	SubmitExtrinsic(meta *types.Metadata, c types.Call, krp signature.KeyringPair) (txHash types.Hash, bn types.BlockNumber, sig types.Signature, err error)
+}
+
+type api struct {
+	getBlockHash            func(uint64) (types.Hash, error)
+	getRuntimeVersionLatest func() (*types.RuntimeVersion, error)
+	getStorageLatest        func(key types.StorageKey, target interface{}) error
+	getClient               func() client.Client
+	getBlockLatest          func() (*types.SignedBlock, error)
+	getMetadataLatest       func() (*types.Metadata, error)
+}
+
+// NewAPI returns a new centrifuge chain api.
+func NewAPI(sapi *gsrpc.SubstrateAPI) API {
+	return api{
+		getBlockHash:            sapi.RPC.Chain.GetBlockHash,
+		getRuntimeVersionLatest: sapi.RPC.State.GetRuntimeVersionLatest,
+		getStorageLatest:        sapi.RPC.State.GetStorageLatest,
+		getClient:               func() client.Client { return sapi.Client },
+		getBlockLatest:          sapi.RPC.Chain.GetBlockLatest,
+		getMetadataLatest:       sapi.RPC.State.GetMetadataLatest,
+	}
+}
+
+func (a api) GetMetadataLatest() (*types.Metadata, error) {
+	return a.getMetadataLatest()
+}
+
+func (a api) SubmitExtrinsic(meta *types.Metadata, c types.Call, krp signature.KeyringPair) (txHash types.Hash, bn types.BlockNumber, sig types.Signature, err error) {
+	ext := types.NewExtrinsic(c)
+	era := types.ExtrinsicEra{IsMortalEra: false}
+
+	genesisHash, err := a.getBlockHash(0)
+	if err != nil {
+		return txHash, bn, sig, err
+	}
+
+	rv, err := a.getRuntimeVersionLatest()
+	if err != nil {
+		return txHash, bn, sig, err
+	}
+
+	key, err := types.CreateStorageKey(meta, "System", "AccountNonce", krp.PublicKey)
```

Comment:
I was wondering that we should probably add a lock to this method so we can ensure that the accountNonce is increased in the pool before sending it over. From my local testing sending multiple txs it is very possible that some of them will fail for this reason, WDYT?

Reply:
@mikiquantum I did actually manually increase the nonce for the xts I sent in parallel, which also led to dropped xts.

Reply:
Yes, I did the same too. I was referring more in the future when we fix the problem we currently have in the centchain. 
At least from the same node (same accounts) we should be able to have a serial sending of txs so we can ensure the nonce increment, either by a lock (at the beginning) or something more complex in the future by keeping a state in the client node. Up to discussion.

Reply:
Makes sense, so we lock in order to prevent 2 threads/workers using the same account with a nonce sequence?

Reply:
Exactly, it is not perfect, since it is possible that the go-centrifuge points to an infura like setup and the pending pool is not shared, it could still fail "silently". But it is a start.

Reply:
Got it. On top of that, xts in pending pool might not increase the nonce - but we need to resend failed xts anyways.

Reply:
hmm, interesting one, I will research on that as soon as I get to the integration testing.

Reply:
> Got it. On top of that, xts in pending pool might not increase the nonce - but we need to resend failed xts anyways.

This is big issue I feel. Wouldn't System.nonce take of this already?

Anyway, I added a lock for each extrinsic. This will ensure only one txn is sent at once.